esp/esp/program/modules/handlers/teachercheckinmodule.py

```diff
@@ -149,13 +150,31 @@ def ajaxteachercheckin(self, request, tl, one, two, module, extra, prog):
         return HttpResponse(json.dumps(json_data), content_type='text/json')
 
     @staticmethod
-    def get_phone(user):
-        """Get the phone number to display for a user."""
-        contact_info = user.getLastProfile().contact_user
+    def get_phones(users):
+        """
+        Given a list or QuerySet of users, create a dictionary that maps user
+        ids to phone numbers for displaying.
+        """
+
         default = '(missing contact info)'
-        if contact_info:
-            return contact_info.phone_cell or contact_info.phone_day or default
-        return default
+
+        # This is an optimized version of doing this for each user:
+
+        #   contact_info = user.getLastProfile().contact_user
+        #   if contact_info:
+        #       return contact_info.phone_cell or contact_info.phone_day or default
+        #   return default
+
+        # Only Postgres supports the following fancy database operation! See
+        # http://stackoverflow.com/a/20129229/3243497 .
+
+        profiles = (RegistrationProfile.objects
+                .filter(user__in=users)
+                .order_by('user__id', '-last_ts')
+                .distinct('user__id')
```

Comment:
Neat trick!  I wonder if we will want to use this elsewhere where we do crazy things to fetch lots of reg profile info.